diagnostics/include/diagnostics.h

```diff
@@ -0,0 +1,82 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+#include <functional>
+#include <map>
+#include <mutex>
+#include <stdexcept>
+
+namespace concord::diagnostics {
+
+// Every component that can return status should register a status handler with the Registrar.
+struct StatusHandler {
+  StatusHandler(std::string name, std::string description, std::function<std::string()> fun)
+      : name(std::move(name)), description(std::move(description)), f(std::move(fun)) {}
+  std::string name;
+  std::string description;
+  std::function<std::string()> f;
+};
+
+class Registrar {
+ public:
+  void registerStatusHandler(StatusHandler handler) {
+    std::lock_guard<std::mutex> guard(mutex_);
+    if (status_handlers_.count(handler.name)) {
+      throw std::invalid_argument(std::string("StatusHandler already exists: ") + handler.name);
+    }
+    status_handlers_.insert({handler.name, handler});
```

Comment:
[Minor]: Just a thought - that operation can be done with a single lookup by calling `insert()` directly and checking the return value to determine whether an insertion took place. I doubt there will be lots of handlers and it shouldn't have a big impact.

Reply:
That's true, but it doesn't have the same behavior. In the code as it currently is written, the first handler is never overwritten. In that case a caller could theoretically handle the exception and try a different name, etc... It's an unlikely problem, but I prefer to keep it this way.

Reply:
You are absolutely right. I should stop doing reviews yearly in the morning...